### ISSUE TITLE
problems with the loss updating

### ISSUE THREAD

**Issue by CarlosNacher** [HuiZhang0812/DiffusionAD#28]
Hi @HuiZhang0812 ,

First of all, thanks a lot for your work and for sharing it here! 

I have realized one thing what I think could be a coding error. In `train.py` file, `lines 136-141` the train losses are added to the epoch total loss:

```
train_loss += loss.item()
tbar.set_description('Epoch:%d, Train loss: %.3f' % (epoch, train_loss))

train_smL1_loss += smL1_loss.item()
train_focal_loss+=5*focal_loss.item()
train_noise_loss+=noise_loss.item()
```
I suppose this was for after, within `lines 144-149`, compute the **mean of those epoch losses** dividing them with the number of steps of the epoch (`i`). **But it seems to have been forgotten, and it is being appended the sum of the epoch loss**.

I think **the following:**
 ```
        if epoch % 10 ==0  and epoch > 0:
            train_loss_list.append(round(train_loss,3))
            train_smL1_loss_list.append(round(train_smL1_loss,3))
            train_focal_loss_list.append(round(train_focal_loss,3))
            train_noise_loss_list.append(round(train_noise_loss,3))
            loss_x_list.append(int(epoch))
```

Has to be **replaced with:**

 ```
        if epoch % 10 ==0  and epoch > 0:
            train_loss_list.append(round(train_loss / (i+1),3))
            train_smL1_loss_list.append(round(train_smL1_losss / (i+1),3))
            train_focal_loss_list.append(round(train_focal_losss / (i+1),3))
            train_noise_loss_list.append(round(train_noise_losss / (i+1),3))
            loss_x_list.append(int(epoch))
```

_(to take advantage of the `i` variable, obviously there are alternative ways to achieve it)._

In addition, these loss lists are not saved or logged anywhere afterwards (I guess this is a decision you are aware of).

Again, thank you very much for your great work in developing this SOTA algorithm and even more for sharing it with the community.

**Comment by CarlosNacher** [comment]
**UPDATE:** Regarding that when one batch has all images with anomalies (`anomaly_label ==1`) the loss computed within the funcion `norm_guided_one_step_denoising` of the `DDPM.py` file sets to **nan**:

`line 359 -> loss = (normal_loss["loss"]+noisier_loss["loss"])[anomaly_label==0].mean()` 

**Adding losses** to the epoch total losses:

```
train_loss += loss.item()
tbar.set_description('Epoch:%d, Train loss: %.3f' % (epoch, train_loss))

train_smL1_loss += smL1_loss.item()
train_focal_loss+=5*focal_loss.item()
train_noise_loss+=noise_loss.item()
```

**leads to the fact that the errors of each epoch that are stored in the lists are always nan** from the time the first nan appears.

So a **possible solution** is adding them only if they are not nan

**Comment by HuiZhang0812** [comment]
Thank you for your suggestion. The code you mentioned in lines 144-149 is designed to observe the variation of the loss, so summation or averaging might both be feasible, it's just a matter of scaling.

**Comment by CarlosNacher** [comment]
> Thank you for your suggestion. The code you mentioned in lines 144-149 is designed to observe the variation of the loss, so summation or averaging might both be feasible, it's just a matter of scaling.

Totally agree that it is a matter of scale. Usually the average is used to make it more representative of "the loss of the epoch". However, I agree that if all you want to do is look at the variation, it doesn't matter what the scale is.

Thank you very much for the clarification and for your huge contribution
